src/algorithms/tonal/nnlschroma.cpp

```diff
@@ -0,0 +1,553 @@
+  vector<vector<Real> >& chromagram = _chromagram.get();
+
+  if (logSpectrum.size() <= 1)
+    throw EssentiaException("NNLSChroma: input vector is empty");
+
+  if (logSpectrum[0].size() != 256) {
+    throw EssentiaException("NNLSChroma: log spectrum size should be 256 but it is ", 
+                            logSpectrum[0].size(), ".");
+  }
+
+  /**  Calculate Tuning
+       calculate tuning from (using the angle of the complex number defined by the 
+       cumulative mean real and imag values)
+  **/
+  Real meanTuningImag = 0;
+  Real meanTuningReal = 0;
+  for (int iBPS = 0; iBPS < nBPS; ++iBPS) {
+    meanTuningReal += meanTuning[iBPS] * _cosvalues[iBPS];
+    meanTuningImag += meanTuning[iBPS] * _sinvalues[iBPS];
+  }
+
+  Real normalisedtuning = atan2(meanTuningImag, meanTuningReal) / (2 * M_PI);
+  int intShift = floor(normalisedtuning * 3);
+  Real RealShift = normalisedtuning * 3 - intShift; // RealShift is a really bad name for this
+
+
+  /** Tune Log-Frequency Spectrogram
+      calculate a tuned log-frequency spectrogram (tunedLogfreqSpectrum): use the tuning estimated above (kinda f0) to 
+      perform linear interpolation on the existing log-frequency spectrogram (kinda f1).
+  **/   
+  Real tempValue = 0;
+
+  tunedLogfreqSpectrum.assign(logSpectrum.size(), vector<Real>(2, 0.f));
+  
+  for (int i = 0; i < (int)logSpectrum.size(); i++) {
+    if (_tuningMode) {
+        intShift = floor(localTuning[i] * 3.f);
+        RealShift = localTuning[i] * 3.f - intShift; // RealShift is a really bad name for this
+    }
+        
+    // Interpolate all inner bins.
+    for (int k = 2; k < (int)logSpectrum[i].size() - 3; ++k) {
+      tempValue = logSpectrum[i][k + intShift] * (1 - RealShift) +
+                  logSpectrum[i][k + intShift + 1] * RealShift;
+      tunedLogfreqSpectrum[i].push_back(tempValue);
+    }
+
+    tunedLogfreqSpectrum[i].push_back(0.0);
+    tunedLogfreqSpectrum[i].push_back(0.0);
+    tunedLogfreqSpectrum[i].push_back(0.0);  // upper edge
+
+    vector<Real> runningmean = SpecialConvolution(tunedLogfreqSpectrum[i], _hw);
+    vector<Real> runningstd;
+
+    // First step: squared values into vector (variance).
+    for (int j = 0; j < nNote; j++) {
+      runningstd.push_back((tunedLogfreqSpectrum[i][j] - runningmean[j]) *
+                           (tunedLogfreqSpectrum[i][j] - runningmean[j]));
+    }
+
+    // Second step convolve.
+    runningstd = SpecialConvolution(runningstd, _hw);
+
+    for (int j = 0; j < nNote; j++) {  
+      runningstd[j] = sqrt(runningstd[j]); // square root to finally have running std
+      if (runningstd[j] > 0) {
+        tunedLogfreqSpectrum[i][j] = (tunedLogfreqSpectrum[i][j] - runningmean[j]) > 0 ?
+          (tunedLogfreqSpectrum[i][j] - runningmean[j]) / pow(runningstd[j], _whitening) : 0;
+      }
+      if (tunedLogfreqSpectrum[i][j] < 0) {
+        E_INFO("ERROR: negative value in logfreq spectrum");
```

Comment:
Should this never happen in practice? We can print this INFO message, but we could also through an exception if that makes sense.